blockdevice/table/gpt/gpt.go

```diff
@@ -270,31 +298,58 @@ func (gpt *GPT) Repair() error {
 	return nil
 }
 
-// Add adds a partition.
+// Add adds a partition to the end of the list.
 func (gpt *GPT) Add(size uint64, setters ...interface{}) (table.Partition, error) {
+	return gpt.InsertAt(len(gpt.partitions), size, setters...)
+}
+
+// InsertAt inserts partition after the partition at the position idx.
+//
+// If idx == 0, it inserts new partition as the first partition, etc., idx == 1 as the second, etc.
+func (gpt *GPT) InsertAt(idx int, size uint64, setters ...interface{}) (table.Partition, error) {
 	opts := partition.NewDefaultOptions(setters...)
 
-	var start, end uint64
-	if len(gpt.partitions) == 0 {
-		start = gpt.header.FirstUsableLBA
-	} else {
-		previous := gpt.partitions[len(gpt.partitions)-1]
-		start = previous.(*partition.Partition).LastLBA + 1
+	// find the minimum and maximum LBAs available
+	var minLBA, maxLBA uint64
+
+	minLBA = gpt.header.FirstUsableLBA
+
+	for i := idx - 1; i >= 0; i-- {
+		if gpt.partitions[i] != nil {
+			minLBA = gpt.partitions[i].(*partition.Partition).LastLBA + 1
+
+			break
+		}
+	}
+
+	maxLBA = gpt.header.LastUsableLBA
+
+	for i := idx; i < len(gpt.partitions); i++ {
+		if gpt.partitions[i] != nil {
+			maxLBA = gpt.partitions[i].(*partition.Partition).FirstLBA - 1
+
+			break
+		}
 	}
 
+	// find partition boundaries
+	var start, end uint64
+
+	start = minLBA
+
 	if opts.MaximumSize {
-		end = gpt.header.LastUsableLBA
+		end = maxLBA
 
-		if end <= start {
+		if end < start {
 			return nil, fmt.Errorf("requested partition with maximum size, but no space available")
```

Comment:
Should we make this a sentinel error?

```
var ErrNoSpaceAvailable = errors.New("no space available")
```

Reply:
I think if we are to do that, it's better to have correct error interface and checks, as sentinel errors don't allow messages to be customized

Reply:
reworked this way

Reply:
Either way works for me. I think as long as a user can check the error type, its good.